src/vs/base/common/event.ts

```diff
@@ -1563,30 +1563,47 @@ export class DynamicListEventMultiplexer<TItem, TEventType> implements IDynamicL
  * // event will only be fired at this point
  * ```
  */
-export class EventBufferer {
+export class EventBufferer<T> {
 
-	private buffers: Function[][] = [];
+	private data: { buffers: Function[]; items: T[] }[] = [];
 
-	wrapEvent<T>(event: Event<T>): Event<T> {
+	wrapEvent<T>(event: Event<T>): Event<T>;
+	wrapEvent<T>(event: Event<T>, reduce: (last: T | undefined, event: T) => T): Event<T>;
+	wrapEvent<O>(event: Event<T>, reduce: (last: O | undefined, event: T) => O, initial: O): Event<O>;
+	wrapEvent<O>(event: Event<T>, reduce?: (last: T | O | undefined, event: T) => T | O, initial?: O): Event<O | T> {
 		return (listener, thisArgs?, disposables?) => {
 			return event(i => {
-				const buffer = this.buffers[this.buffers.length - 1];
-
-				if (buffer) {
-					buffer.push(() => listener.call(thisArgs, i));
+				const data = this.data[this.data.length - 1];
+				if (reduce) {
+					if (data) {
+						if (data.buffers.length === 0) {
+							if (initial) {
+								data.buffers.push(() => listener.call(thisArgs, data.items.reduce(reduce, initial)));
+							} else {
+								data.buffers.push(() => listener.call(thisArgs, data.items.reduce(reduce as (last: T | undefined, event: T) => T)));
```

Comment:
The `reduce` operation will run for each listener when events get flushed. Any way we can just run it once and give the result to each listener instead?

Reply:
It should only run once because of the `data.buffers.length` check... but let me make it a bit more diliberate.

Reply:
How's this?